internal/apicodec/codec_v2.go

```diff
@@ -0,0 +1,789 @@
+package apicodec
+
+import (
+	"bytes"
+
+	"github.com/pingcap/kvproto/pkg/errorpb"
+	"github.com/pingcap/kvproto/pkg/kvrpcpb"
+	"github.com/pkg/errors"
+	"github.com/tikv/client-go/v2/tikvrpc"
+)
+
+var (
+	// APIV2RawKeyPrefix is prefix of raw key in API V2.
+	APIV2RawKeyPrefix = []byte{'r', 0, 0, 0}
+
+	// APIV2RawEndKey is max key of raw key in API V2.
+	APIV2RawEndKey = []byte{'r', 0, 0, 1}
+
+	// APIV2TxnKeyPrefix is prefix of txn key in API V2.
+	APIV2TxnKeyPrefix = []byte{'x', 0, 0, 0}
+
+	// APIV2TxnEndKey is max key of txn key in API V2.
+	APIV2TxnEndKey = []byte{'x', 0, 0, 1}
+)
+
+// codecV2 is used to encode/decode keys and request into APIv2 format.
+type codecV2 struct {
+	prefix   []byte
+	endKey   []byte
+	memCodec memCodec
+}
+
+// NewCodecV2 returns a codec that can be used to encode/decode
+// keys and requests to and from APIv2 format.
+func NewCodecV2(mode Mode) Codec {
+	// Region keys in CodecV2 are always encoded in memory comparable form.
+	codec := &codecV2{memCodec: &memComparableCodec{}}
+	switch mode {
+	case ModeRaw:
+		codec.prefix = APIV2RawKeyPrefix
+		codec.endKey = APIV2RawEndKey
+	case ModeTxn:
+		codec.prefix = APIV2TxnKeyPrefix
+		codec.endKey = APIV2TxnEndKey
```

Comment:
The prefix and end key should not be hard-coded here. We could derive the end key by the prefix: `be_bytes(end) == be_bytes(be(prefix) + 1)`. We might provide a factory to construct the Codec by another prefix while we could acquire keyspace meta from PD.